doc/murmur3sum.1
.TH MURMUR3SUM 1 "October 2022" "murmur3sum 1.x" "User Commands"
.SH NAME
murmur3sum \- Compute or verify the murmur3 hash of files
.SH SYNOPSIS
.B murmur3sum
[\fI\-\-OPTION\fR] [--] [<file1> <file2> ... <fileN>] [> <checksum-file>]
.PP
.B murmur3sum --check
[\fI\-\-OPTION\fR] [<checksum-file>]
.SH DESCRIPTION
The \fBmurmur3sum(1)\fR tool is similar to the \fBmd5sum(1)\fR tool. It
reads files and computes a hash based on the Murmur3 hashing scheme.

Without the --check (or -c) option, the specified files are read and their
hash computed and printed to stdout following the md5sum format:

    00000000000000000000000000000000  some/file/name.here

where the 0's are replaced by the hash and the "some/file/name.here" is
the path as specified on the command line.

The special "-" filename represents standard input. If no filename is
specified, then the tool automatically uses "-". In other words, you can
pipe data through this tool like so:

    cat file | murmur3sum --no-filename

The --no-filename is probably a good idea in this case, otherwise it uses
"-" as the filename which doesn't mean much.

.SH "VERIFY FILES"
The tool can be used to verify a set of files.

First, you create a checksum file like so:

    murmur3sum "file1" "file2" ... "fileN" > checksums.murmur3

Later, you can verify that the files have not changed as follow:

    murmur3sum --check checksums.murmur3

The tool makes use of the hashes and filenames found inside the
checksums.murmur3 file to verify the current hashes of the files on disk.
If they changed, an error is emitted. If they are equal, it shows the filename
followed by "OK" (which you can --quiet if you do not need to see the results).

In a script, you can use the "if" instruction like so:

    if ! murmur3sum --quiet --check checksums.murmur3
    then
        # detected an error
.br
	   ...handle error...
    fi

.SH "COMMAND LINE OPTIONS"
The following describes all the supported command line options. The
special "--" \fIoption\fR is used to stop the parsing of options and
forcibly view all the remaining command line arguments as filenames.

.TP
\fB\-b\fR, \fB--binary\fR
Read the input files in binary. This is always the case. We do not support
a text mode.

.TP
\fB\-c\fR, \fB--check\fR
Check the hash of files as saved in a text which is read by \fBmurmur3sum(1)\fR.
Each line is expected to be defined as above (hash, two spaces, the filename).
It verifies that the hash is a valid string of hexadecimal digits. If so, it
uses the filename to read the file and compute its hash. If that computed
hash is equal to the hash found in the checksum file, then the filename is
written in stdout followed by "OK". Otherwise an error is printed. If one
or more errors occur, then the tool exits with 1.

.TP
\fB\-h\fR, \fB\-\-help\fR
Print a brief document about the tool usage, then exit.

.TP
\fB\-\-basename\fR
Output the basename of each file instead of their complete path. This can be
useful for files being installed in a known location later.

.TP
\fB\-\-license\fR, \fB\-\-version\fR
Display the version and license information of the murmur3sum tool.

.TP
\fB\-\-no-filename\fR
Only output the hash. This is particularly useful when computing a hash
for one file only and the filename is really not required in the output.

.TP
\fB\-t\fR, \fB\-\-text\fR
This command line option is ignored. We always read the files in binary.

.TP
\fB\-\-quiet\fR
Do not display the "filename: OK" message when checking files.

.SH AUTHOR
Written by Alexis Wilke <alexis@m2osw.com>.
.SH "REPORTING BUGS"
Report bugs to <https://github.com/m2osw/murmur3/issues>.
.br
iplock home page: <https://snapwebsites.org/project/murmur3>.
.SH COPYRIGHT
Copyright \(co 2022-2025  Made to Order Software Corp.  All Rights Reserved
.br
License: GPLv3
.br
This is free software: you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR md5sum (1)


debian/changelog
murmur3 (1.0.7.0~jammy) jammy; urgency=high

  * Removing boost as a dependency.
  * Fixed version date in murmur3sum command line tool.
  * Added a missing #include <cstdint>
  * Added missing binaries to the .install file.
  * Install the man page.
  * Added more documentation.
  * Added a logo for the README.md.
  * Updated the compat to 15.
  * Actually install the stream.h header.
  * Bumped copyright notice to 2023.
  * Removed boost as a dependency.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 04 Mar 2023 09:45:08 -0800

murmur3 (1.0.6.0~bionic) bionic; urgency=high

  * Properly renamed the packages.

 -- Alexis Wilke <alexis@m2osw.com>  Wed, 02 Nov 2022 12:47:33 -0700

murmur3 (1.0.5.0~bionic) bionic; urgency=high

  * Changed the project name from libmurmur3 to just murmur3.
  * Added the murmur3sum tool (similar to md5sum with murmur3).
  * Added a manual page for the murmur3sum tool.
  * Added support for a streaming version in C++.
  * Converted the test to snapcatch2 so we can get coverage data.
  * Fixed seeding of 64 bit C version, it supports 64 or 128 bits now.
  * Added many tests to cover all the new code.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 29 Oct 2022 14:01:03 -0700

murmur3 (1.0.4.6~bionic) bionic; urgency=high

  * Changed compat to v10.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 19 May 2022 20:32:43 -0700

murmur3 (1.0.4.5~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 19 May 2022 10:34:40 -0700

murmur3 (1.0.4.4~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 29 Aug 2021 12:14:49 -0700

murmur3 (1.0.4.3~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 24 Aug 2021 14:05:06 -0700

murmur3 (1.0.4.2~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 04 Jun 2021 17:47:35 -0700

murmur3 (1.0.4.1~bionic) bionic; urgency=low

  * Bumped version to recompile.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 15 May 2021 10:01:50 -0700

murmur3 (1.0.4.0~bionic) bionic; urgency=low

  * Changed the version to bionic.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 13 Nov 2020 20:46:54 -0800

murmur3 (1.0.3.1~xenial) xenial; urgency=low

  * Bump version to get another build request.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 26 May 2020 11:20:49 -0800

murmur3 (1.0.3.0~xenial) xenial; urgency=low

  * Make the fallthrough attribute optional (requires gcc 7+).

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 25 May 2020 20:54:58 -0800

murmur3 (1.0.2.0~xenial) xenial; urgency=low

  * Added the fallthrough attribute so it compiles under Ubuntu 18.04.
  * Updated the compat version to 9.
  * Fixed the murmur3 names and content.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 26 Jan 2020 20:03:39 -0800

murmur3 (1.0.1.0~xenial) xenial; urgency=low

  * Fixed the build dependencies.
  * Added a TODO.md file.
  * Fixed the copyright notice.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 19 Oct 2019 02:14:06 -0800

murmur3 (1.0.0.0~xenial) xenial; urgency=low

  * Initial release

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 19 Oct 2019 00:48:35 -0800


tools/murmur3sum.cpp
// Copyright (c) 2012-2025  Made to Order Software Corp.  All Rights Reserved
//
// https://snapwebsites.org/project/murmur3
// contact@m2osw.com
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.


// murmur3
//
#include    <murmur3/stream.h>
#include    <murmur3/version.h>


// snapdev
//
#include    <snapdev/hexadecimal_string.h>


// C++
//
#include    <iostream>
#include    <vector>


// C
//
#include    <limits.h>
#include    <stdio.h>
#include    <string.h>


namespace
{

std::string                     g_progname = std::string();
bool                            g_basename = false;
bool                            g_no_filename = false;
bool                            g_check = false;
bool                            g_quiet = false;
std::vector<std::string>        g_filenames = {};
std::size_t                     g_error = 0;


void usage()
{
    std::cout << "Usage: " << g_progname << " [--<opts>] [--] [-] [filename]\n";
    std::cout << "where --opts is one or more of:\n";
    std::cout << "   -b | --binary   read file in binary\n";
    std::cout << "   -c | --check    the input files are lists of checksum, read the corresponding file and verify the hash\n";
    std::cout << "   -h | --help     print out this help screen\n";
    std::cout << "   --basename      only print basename of filename\n";
    std::cout << "   --license       print out version and license information\n";
    std::cout << "   --no-filename   output the hash without the filename\n";
    std::cout << "   -t | --text     read file in text (this is ignored, we always read in binary)\n";
    std::cout << "   --quiet         check quietly (do not print OK for each file)\n";
    std::cout << "   --version       print out version and license information\n";
}


void generate_hashes()
{
    for(auto const & f : g_filenames)
    {
        FILE * in(nullptr);
        if(f == "-")
        {
            in = stdin;
        }
        else
        {
            in = fopen(f.c_str(), "rb");
            if(in == nullptr)
            {
                int const e(errno);
                std::cerr
                    << "error:"
                    << g_progname
                    << ": could not open \""
                    << f
                    << "\" for reading: "
                    << strerror(e)
                    << "\n";
                ++g_error;
                continue;
            }
        }
        bool valid(true);
        murmur3::stream sum(0);  // no seeding (add option to command line?)
        for(;;)
        {
            char buffer[4096];
            std::size_t const size(fread(buffer, sizeof(buffer[0]), sizeof(buffer) / sizeof(buffer[0]), in));
            if(size == 0)
            {
                if(ferror(in))
                {
                    int const e(errno);
                    std::cerr
                        << "error:"
                        << g_progname
                        << ": an I/O error occurred while reading \""
                        << f
                        << "\": "
                        << strerror(e)
                        << "\n";
                    ++g_error;
                    valid = false;
                }
                break;
            }
            sum.add_data(buffer, size);
        }
        if(valid)
        {
            std::cout << sum.flush().to_string();
            if(!g_no_filename)
            {
                std::string::size_type const last_slash((g_basename ? f.rfind('/') : std::string::npos) + 1);
                std::cout << "  " << f.substr(last_slash);
            }
            std::cout << "\n";
        }
        if(f != "-")
        {
            fclose(in);
        }
    }
}


void check_hashes()
{
    for(auto const & f : g_filenames)
    {
        FILE * in(nullptr);
        if(f == "-")
        {
            in = stdin;
        }
        else
        {
            in = fopen(f.c_str(), "rb");
            if(in == nullptr)
            {
                int const e(errno);
                std::cerr
                    << "error:"
                    << g_progname
                    << ": could not open \""
                    << f
                    << "\" for reading: "
                    << strerror(e)
                    << "\n";
                ++g_error;
                continue;
            }
        }
        char line[murmur3::HASH_SIZE * 2 + 2 + PATH_MAX + 2]; // try to allow for a long filename
        line[sizeof(line) - 1] = '\0';
        int l(0);
        while(fgets(line, sizeof(line), in) != nullptr)
        {
            ++l;

            // make sure the input line is valid (not empty, not too long)
            //
            std::size_t const len(strlen(line));
            if(len == 0)
            {
                std::cerr
                    << "error: "
                    << g_progname
                    << ':'
                    << l
                    << ": could not read line of data (empty).\n";
                ++g_error;
                break;
            }
            if(line[len - 1] != '\n')
            {
                std::cerr
                    << "error:"
                    << g_progname
                    << ':'
                    << l
                    << ": line too long.\n";
                ++g_error;
                break;
            }
            if(len == 1)
            {
                // ignore empty lines
                //
                continue;
            }
            if(line[0] == '#')
            {
                // ignore lines commented out
                //
                continue;
            }

            // extract the expected hash from the murmur3 file
            //
            std::string const hex(line, 32);
            murmur3::hash expected_hash;
            try
            {
                expected_hash.from_string(hex);
            }
            catch(snapdev::hexadecimal_string_invalid_parameter const & e)
            {
                std::cerr
                    << "error:"
                    << g_progname
                    << ':'
                    << l
                    << ": hash ("
                    << hex
                    << ") could not be decoded: "
                    << e.what()
                    << "\n";
                ++g_error;
                continue;
            }

            // retrieve the filename
            //
            // TBD: do we want to verify the two spaces between hash & filename?
            //
            std::string const filename(line + 34, len - 34 - 1);    // -1 for the '\n'
            murmur3::hash const computed_hash(murmur3::sum(filename));

            if(expected_hash != computed_hash)
            {
                std::cerr
                    << "error:"
                    << g_progname
                    << ':'
                    << l
                    << ": hash found in checksum file ("
                    << hex
                    << ") does not match hash ("
                    << computed_hash.to_string()
                    << ") computed from file \""
                    << filename
                    << "\".\n";
                ++g_error;
                continue;
            }

            if(!g_quiet)
            {
                std::cout << filename << ": OK\n";
            }
        }
        if(f != "-")
        {
            fclose(in);
        }
    }
}


} // no name namespace


int main(int argc, char * argv[])
{
    g_progname = argv[0];
    std::string::size_type const slash(g_progname.rfind('/'));
    if(slash != std::string::npos)
    {
        g_progname = g_progname.substr(slash + 1);
    }

    bool filename_only(false);
    for(int i(1); i < argc; ++i)
    {
        if(!filename_only && argv[i][0] == '-')
        {
            switch(argv[i][1])
            {
            case '-':   // long option
                if(argv[i][2] == '\0')
                {
                    // found "--", only filenames after that
                    filename_only = true;
                    break;
                }
                if(strcmp(argv[i] + 2, "help") == 0)
                {
                    usage();
                    return 1;
                }
                if(strcmp(argv[i] + 2, "basename") == 0)
                {
                    g_basename = true;
                }
                else if(strcmp(argv[i] + 2, "binary") == 0)
                {
                    ; // ignore
                }
                else if(strcmp(argv[i] + 2, "check") == 0)
                {
                    g_check = true;
                }
                else if(strcmp(argv[i] + 2, "no-filename") == 0)
                {
                    g_no_filename = true;
                }
                else if(strcmp(argv[i] + 2, "quiet") == 0)
                {
                    g_quiet = true;
                }
                else if(strcmp(argv[i] + 2, "text") == 0)
                {
                    ; // ignore
                }
                else if(strcmp(argv[i] + 2, "license") == 0
                     || strcmp(argv[i] + 2, "version") == 0)
                {
                    std::cout
                        << g_progname
                        << " (Snap! C++ Project) "
                        << MURMUR3_VERSION_STRING
                        << '\n'
                        << "Copyright (C) 2022-" << UTC_BUILD_YEAR << " Made to Order Software Corporation\n"
                           "License GPVv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n"
                           "This is free software: you are free to change and redistribute it.\n"
                           "There is NO WARRANTY, to the extent permitted by law.\n"
                           "\n"
                           "Written by Alexis Wilke, Peter Scott, and A. Appleby.\n";
                    return 0;
                }
                break;

            case 'h':
                usage();
                return 1;

            case 'c':
                g_check = true;
                break;

            case 'b': // binary
            case 't': // text
                // just ignore
                break;

            }
        }
        else
        {
            g_filenames.push_back(argv[i]);
        }
    }

    // if no filename was specified, use stdin
    //
    if(g_filenames.empty())
    {
        g_filenames.push_back("-");
    }

    if(g_check)
    {
        check_hashes();
    }
    else
    {
        generate_hashes();
    }

    if(g_error != 0)
    {
        std::cerr << "error:"
            << g_progname
            << ": found "
            << g_error
            << " errors.\n";
        return 1;
    }

    return 0;
}

// vim: ts=4 sw=4 et
